rfcs/20200928-keras-cv-single-stage-2d-object-detection.md

```diff
@@ -0,0 +1,431 @@
+4. metrics utilities such as COCO Evaluator, or visualization utils.
+
+Data augmentation will be included as a separate RFC that handles a
+broader context than object detection.
+
+Model backbone and detection heads are model-specific, we anticipate them to be analyzed and proposed in 
+`keras.applications` for heavily used patterns, however the user can build them easily using Keras.
+
+#### Training
+
+Case where a user want to train from scratch:
+
+```python
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_cv
+
+# Considering a COCO dataset
+coco_dataset = tfds.load('coco/2017')
+train_ds, eval_ds = coco_dataset['train'], coco_dataset['validation']
+
+def preprocess(features):
+  image, gt_boxes, gt_labels = features['image'], features['objects']['bbox'], features['objects']['label']
+  # preprocess image, gt_boxes, gt_labels, such as flip, resize, and padding, and reserve 0 for background label.
+  return image, gt_boxes, gt_labels
+
+anchor_generator = keras_cv.ops.AnchorGenerator(anchor_sizes, scales, aspect_ratios, strides)
+similarity_calculator = keras_cv.ops.IOUSimilarity()
+box_matcher = keras_cv.ops.BoxMatcher(positive_threshold, negative_threshold)
+target_gather = keras_cv.ops.TargetGather()
+box_coder = keras_cv.ops.BoxCoder(offset='sigmoid')
+
+def encode_label(image, gt_boxes, gt_labels):
+  anchor_boxes = anchor_generator(image_size)
+  iou = similarity_calculator(gt_boxes, anchor_boxes)
+  match_indices, match_indicators = box_matcher(iou)
+
+  mask = tf.less_equal(match_indicators, 0)
+  class_mask = tf.expand_dims(mask, -1)
+  box_mask = tf.tile(class_mask, [1, 4])
+
+  class_targets = target_gather(gt_labels, match_indices, class_mask, -1)
+  box_targets = target_gather(gt_boxes, match_indices, box_mask, 0.0)
+  box_targets = box_coder.encode(box_targets, anchor_boxes)
+
+  weights = tf.squeeze(tf.ones_like(gt_labels), axis=-1)
+  ignore_mask = tf.equal(match_indicators, -2)
+  class_weights = target_gather(weights, match_indices, ignore_mask, 0.0)
+  box_weights = target_gather(weights, match_indices, mask, 0.0)
+
+  return (image, {'classification': class_targets, 'regression': box_targets},
+          {'classification': class_weights, 'regression': box_weights})
+
+class RetinaNet(tf.keras.Model):
+  # includes backbone and feature pyramid head.
+  def __init__(self):
+    # self.backbone = Model Backbone that returns dict of feature map
+    # self.fpn = Feature Pyramid Heads that
+    # self.head = classification and regression heads
+  
+  def call(self, image, training=None):
+    feature_map = self.backbone(image, training)
+    feature_map = self.fpn(feature_map, training)
+    class_scores, boxes = self.head(feature_map, training)
+    return {'classification': class_scores, 'regression': boxes}
+
+transformed_train_ds = train_ds.map(preprocess).map(encode_label).batch(128).shuffle(1024)
+transformed_eval_ds = eval_ds.map(preprocess).map(encode_label).batch(128)
```

Comment:
Should we have a preprocessing layer for the image to make the model more encapsulated?

Reply:
Data augmentation is not part of this proposal.

Reply:
Data augmentation is not part of this proposal. But yes, it is part of the roadmap.